src/Microsoft.ML.Data/Data/SchemaDefinition.cs

```diff
@@ -349,98 +340,132 @@ public static SchemaDefinition Create(Type userType, Direction direction = Direc
-                // Disallow duplicate names, because the field enumeration order is not actually
-                // well defined, so we are not guaranteed to have consistent "hiding" from run to
-                // run, across different .NET versions.
-                if (!colNames.Add(name))
-                    throw Contracts.ExceptParam(nameof(userType), "Duplicate column name '{0}' detected, this is disallowed", name);
+            if (!CheckMemberInfo(memberInfo))
+                return false;
+
+            customAttributes = memberInfo.GetCustomAttributes();
+            var customTypeAttributes = customAttributes.Where(x => x is DataViewTypeAttribute);
+            if (customTypeAttributes.Count() > 1)
+                throw Contracts.ExceptParam(nameof(userType), "Member {0} cannot be marked with multiple attributes, {1}, derived from {2}.",
+                    memberInfo.Name, customTypeAttributes, typeof(DataViewTypeAttribute));
+            else if (customTypeAttributes.Count() == 1)
+            {
+                var customTypeAttribute = (DataViewTypeAttribute)customTypeAttributes.First();
+                customTypeAttribute.Register();
+            }
 
-                InternalSchemaDefinition.GetVectorAndItemType(memberInfo, out bool isVector, out Type dataType);
+            var mappingNameAttr = memberInfo.GetCustomAttribute<ColumnNameAttribute>();
+            name = mappingNameAttr?.Name ?? memberInfo.Name;
+            // Disallow duplicate names, because the field enumeration order is not actually
+            // well defined, so we are not guaranteed to have consistent "hiding" from run to
+            // run, across different .NET versions.
+            if (!colNames.Add(name))
+                throw Contracts.ExceptParam(nameof(userType), "Duplicate column name '{0}' detected, this is disallowed", name);
 
-                // Get the column type.
-                DataViewType columnType;
-                if (!DataViewTypeManager.Knows(dataType, customAttributes))
+            return true;
+        }
+
+        /// <summary>
+        /// Create a schema definition by enumerating all public fields of the given type.
+        /// </summary>
+        /// <param name="userType">The type to base the schema on.</param>
+        /// <param name="direction">Accept fields and properties based on their direction.</param>
+        /// <returns>The generated schema definition.</returns>
+        public static SchemaDefinition Create(Type userType, Direction direction = Direction.Both)
+        {
+            var memberInfos = GetMemberInfos(userType, direction);
+
+            SchemaDefinition cols = new SchemaDefinition();
+            HashSet<string> colNames = new HashSet<string>();
+
+            foreach (var memberInfo in memberInfos)
+            {
+                if (ValidateMemberInfo(memberInfo, userType, colNames, out string name, out IEnumerable<Attribute> customAttributes))
```

Comment:
Suggestion, if possible, move the CheckMemberInfo call outside of ValidateMemberInfo and put it in here:
```
foreach(var memberInfo in memberInfos)
{
   if(!CheckMemberInfo)
        continue

   ValidateMemberInfo(... out name, out customAttributes)
 .....
}
```

I'd think it would be more readable, and also now you won't need to return a bool from the `ValidateMemberInfo`.